### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by torbacz**
## Motivation and Context

- Resolves nanoFramework/Home#1130
- Resolves nanoFramework/Home#1116

## How Has This Been Tested?

- Unit tests (via pipeline build)
- Benchmarks (on ESP32)

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by torbacz** [comment]
Performance results
First 3 colums are related to old code (from main branch)
4-6 Current

```text
| MethodName         | ItterationCount | Mean   | Min    | Max    | Mean   | Min    | Max    |
| --------------------------------------------------------------- |------------------------- |
| IntArray           | 5               | 254 ms | 250 ms | 260 ms | 334 ms | 330 ms | 340 ms |
| ShortArray         | 5               | 246 ms | 240 ms | 250 ms | 312 ms | 310 ms | 320 ms |
| String             | 5               | 86 ms  | 80 ms  | 90 ms  | 78 ms  | 70 ms  | 80 ms  |
| NestedClass        | 5               | 166 ms | 160 ms | 170 ms | 166 ms | 160 ms | 170 ms |
| ComplexObject      | 5               | 286 ms | 280 ms | 290 ms | 290 ms | 290 ms | 290 ms |
| ComplexArrayObject | 5               | 496 ms | 490 ms | 500 ms | 502 ms | 500 ms | 510 ms |
| ArrayList          | 5               | 520 ms | 520 ms | 520 ms | 536 ms | 530 ms | 550 ms |

Console export: ValueTypesDeserializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean   | Min    | Max    | Mean   | Min    | Max    |
| ------------------------------------------------------- |------------------------- |
| Short      | 5               | 198 ms | 190 ms | 200 ms | 194 ms | 190 ms | 200 ms |
| TimeSpanT  | 5               | 316 ms | 310 ms | 320 ms | 312 ms | 310 ms | 320 ms |
| Float      | 5               | 208 ms | 200 ms | 210 ms | 204 ms | 200 ms | 210 ms |
| Double     | 5               | 216 ms | 210 ms | 230 ms | 216 ms | 210 ms | 220 ms |
| DateTimeT  | 5               | 264 ms | 260 ms | 270 ms | 258 ms | 250 ms | 270 ms |
| Long       | 5               | 220 ms | 210 ms | 230 ms | 220 ms | 210 ms | 230 ms |
| EnumBoxed  | 5               | 272 ms | 260 ms | 280 ms | 258 ms | 250 ms | 270 ms |

Console export: ReferenceTypesSerializationBenchmark benchmark class.
| MethodName                    | ItterationCount | Mean   | Min    | Max    | Mean   | Min    | Max    |
| -------------------------------------------------------------------------- |------------------------- |
| IntArray                      | 5               | 698 ms | 690 ms | 710 ms | 644 ms | 640 ms | 650 ms |
| ArrayList                     | 5               | 972 ms | 960 ms | 980 ms | 942 ms | 930 ms | 950 ms |
| ComplexObjectAzureTwinPayload | 5               | 472 ms | 470 ms | 480 ms | 458 ms | 450 ms | 460 ms |
| ShortArray                    | 5               | 762 ms | 750 ms | 780 ms | 634 ms | 630 ms | 640 ms |
| String                        | 5               | 166 ms | 160 ms | 170 ms | 162 ms | 160 ms | 170 ms |
| NestedClass                   | 5               | 170 ms | 160 ms | 180 ms | 156 ms | 150 ms | 160 ms |
| ComplexArrayObject            | 5               | 638 ms | 630 ms | 640 ms | 588 ms | 580 ms | 590 ms |

Console export: ValueTypesSerializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean    | Min      | Max      | Mean    | Min      | Max      |
| ------------------------------------------------------------ |------------------------------ |
| Short      | 5               | 222 ms  | 220 ms   | 230 ms   | 254 ms  | 240 ms   | 270 ms   |
| TimeSpanT  | 5               | 254 ms  | 250 ms   | 260 ms   | 152 ms  | 140 ms   | 160 ms   |
| Float      | 5               | 280 ms  | 270 ms   | 290 ms   | 230 ms  | 230 ms   | 230 ms   |
| Double     | 5               | 276 ms  | 270 ms   | 280 ms   | 236 ms  | 230 ms   | 240 ms   |
| DateTimeT  | 5               | 1412 ms | 1,410 ms | 1,420 ms | 1464 ms | 1,460 ms | 1,470 ms |
| Long       | 5               | 234 ms  | 230 ms   | 240 ms   | 302 ms  | 300 ms   | 310 ms   |
| Enum       | 5               | 286 ms  | 280 ms   | 290 ms   | 378 ms  | 370 ms   | 380 ms   |
| EnumBoxed  | 5               | 892 ms  | 880 ms   | 910 ms   | 1068 ms | 1,060 ms | 1,070 ms |
```

File: nanoFramework.Json/JsonConvert.cs [hunk 1]
```diff
@@ -165,50 +165,72 @@ private static object ConvertToType(Type sourceType, Type targetType, object val
...
-                case nameof(TimeSpan):
-                    return TimeSpanExtensions.TryConvertFromString(value.ToString());
+            if (targetType == typeof(string))
+            {
+                return value.ToString();
+            }
 
-                default:
-                    return value;
+            if (targetType == typeof(string))
```
Comment (torbacz) [inline nanoFramework.Json/JsonConvert.cs:228]:
```suggestion
            if (targetType == typeof(TimeSpan))
```

**Comment by josesimoes** [comment]
On a 1st glance: the performance gains aren't _that_ huge and some have increased substantially ... which is curious... 🤔

**Comment by torbacz** [comment]
> On a 1st glance: the performance gains aren't _that_ huge and some have increased substantially ... which is curious... 🤔

That's true, but as long as serialization is in one method it's very hard to find why it's behaving like that. I'm trying to find a way to refactor this class to be more like independent components, which would be easier to test and benchmark.

**Comment by torbacz** [comment]
@josesimoes 
I've isolate only "ConvertTo" method and also created benchmarks. Checked 3 approaches:
1. Switch with nameof (ConvertTo)
2. If with typeof (ConvertToIf)
3. Hashtable with classes as converters
image

Results from benchmarking only this method are amazing!
image

Few things I need to address before pushing changes, but looks very promising. 
Also, It will resolve https://github.com/nanoframework/Home/issues/1116

**Comment by josesimoes** [comment]
@torbacz I don't have the benchmark numbers, but I trust you when you' say it looks promising! 😄 
Thanks for the effort you're putting on this. This library it's clearly heading towards perfection. 😉

**Comment by torbacz** [comment]
@josesimoes 
No idea why you can't see the screen shot I've posted. 

Results:
Mean value of 25 runs, 1200 conversions in each batch
CovertToType (old) : **2700 ms**
ConvertToTypeIf (suggested): 2067 ms
ConvertToTypeHashTable: **846 ms**

**Comment by josesimoes** [comment]
@torbacz now I understand what I'm seeing! 😅  Thanks for the extra details.

**Comment by torbacz** [comment]
Current results
```text
Console export: ReferenceTypesDeserializationBenchmark benchmark class.
| MethodName         | ItterationCount | MeanOld| MeanNew|
| -------------------------------------------------------|
| IntArray           | 5               | 254 ms | 168 ms |
| ShortArray         | 5               | 246 ms | 166 ms |
| String             | 5               | 86 ms  | 72 ms  |
| NestedClass        | 5               | 166 ms | 144 ms |
| ComplexObject      | 5               | 286 ms | 234 ms |
| ComplexArrayObject | 5               | 496 ms | 420 ms |
| ArrayList          | 5               | 520 ms | 436 ms |


Console export: ValueTypesDeserializationBenchmark benchmark class.
| MethodName | ItterationCount | MeanOld| MeanNew|
| --------------------------------------- -------- --------
| Short      | 5               | 198 ms | 180 ms | 
| TimeSpanT  | 5               | 316 ms | 294 ms |
| Float      | 5               | 208 ms | 192 ms |
| Double     | 5               | 216 ms | 204 ms |
| DateTimeT  | 5               | 264 ms | 244 ms | 
| Long       | 5               | 220 ms | 204 ms | 
| EnumBoxed  | 5               | 272 ms | 254 ms | 

Console export: ReferenceTypesSerializationBenchmark benchmark class.
| MethodName                    | Mean   | MeanNew|
| -------------------------------------------------
| IntArray                      | 698 ms | 626 ms |
| ArrayList                     | 972 ms | 934 ms |
| ComplexObjectAzureTwinPayload | 472 ms | 446 ms |
| ShortArray                    | 762 ms | 744 ms |
| String                        | 166 ms | 160 ms |
| NestedClass                   | 170 ms | 152 ms |
| ComplexArrayObject            | 638 ms | 686 ms |


Console export: ValueTypesSerializationBenchmark benchmark class.
| MethodName | ItterationCount | MeanOld | MeanNew |
| ---------------------------------------- ---------
| Short      | 5               | 222 ms  | 166 ms  |
| TimeSpanT  | 5               | 254 ms  | 160 ms  |
| Float      | 5               | 280 ms  | 202 ms  |
| Double     | 5               | 276 ms  | 184 ms  |
| DateTimeT  | 5               | 1412 ms | 1302 ms |
| Long       | 5               | 234 ms  | 168 ms  |
| Enum       | 5               | 286 ms  | 244 ms  |
| EnumBoxed  | 5               | 892 ms  | 796 ms  |
```

**Comment by Ellerbach** [comment]
I love the progress. For the last Sonar issue, we can ignore as the random generator is in the benchmark tool and we don't have a create pattern for it.

**Comment by torbacz** [comment]
DateTimeExtensions.IsNumeric improvment

IsNumeric    -> Old
IsNumericParse -> New

```text
| MethodName     | ItterationCount | Mean   | Min    | Max    |
| ----------------------------------------------------------- |
| IsNumeric      | 5               | 712 ms | 710 ms | 720 ms |
| IsNumericParse | 5               | 134 ms | 130 ms | 140 ms |
| IsNumericTry   | 5               | 238 ms | 230 ms | 240 ms |
```

**Comment by torbacz** [comment]
Almost there, something wrong with "DataRow" unit tests. It's last thing to fix.

File: nanoFramework.Json/Converters/DateTimeConverter.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            return "\"" + DateTimeExtensions.ToIso8601((DateTime)value) + "\"";
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        public object ToType(object value)
+        {
+            // Not sure hwo this one should work
```
Comment (Ellerbach) [inline nanoFramework.Json/Converters/DateTimeConverter.cs:21]: well, when is this one used? :-)

Comment (torbacz) [reply nanoFramework.Json/Converters/DateTimeConverter.cs:21]: Not sure if it's used at all. "ToType" method should be used when converting from json to object, but looks like DateTime never hits that. 
I left that not implemented yet all tests are passing... It's something for next PR, this one is already too big...

Comment (torbacz) [reply nanoFramework.Json/Converters/DateTimeConverter.cs:21]: The same situation with bool converter.

**Comment by torbacz** [comment]
Finally, It's ready. There are a lot of changes, because fixing one class for Sonar triggered more checks... 

Still something wrong with AZDO test results - data rows tests are not visible as "passed" in test raport. I'm pretty sure it's not related to JSON changes, because tests are displayed in console as "passed"

File: nanoFramework.Json/Converters/DateTimeConverter.cs (hunk 2 above)
Comment (Ellerbach) [reply nanoFramework.Json/Converters/DateTimeConverter.cs:21]: OK. Good like this then.

**Review by josesimoes — APPROVED** [review]
Clear improvements in performance throughout all the code! Pretty impressive,
And so great to see all covered by Unit Tests in such detail.
Looks **great** to me! 💯 🥇 👏🏻

**Comment by josesimoes** [comment]
@torbacz I've cleared Sonarcloud warnings about Random. All good now.
As for the test results, there a a bunch of skipped tests that shouldn't be... 🤔 I'm looking at the test framework output to improve the output about the tests.

**Comment by josesimoes** [comment]
@torbacz can I push to your branch the fixes for the unit test running locally and a couple of improvements?

**Comment by torbacz** [comment]
@josesimoes Sure

**Comment by torbacz** [comment]
@josesimoes Just one more thing, I want to run benchmarks after all changes and compare it to baseline. Tomorrow it should be ready to merge. I'll let you know.

File: nanoFramework.Json/TimeExtensions.cs [hunk 3]
```diff
@@ -127,16 +127,14 @@ public static string ToIso8601(DateTime dt)
         /// </summary>
         /// <param name="str"></param>
         /// <returns></returns>
-        private static bool IsNumeric(string str)
+        public static bool IsNumeric(string str)
```
Comment (torbacz) [inline nanoFramework.Json/TimeExtensions.cs:130]: TODO: make private again.

File: nanoFramework.Json/TimeExtensions.cs [hunk 4]
```diff
@@ -127,16 +127,14 @@ public static string ToIso8601(DateTime dt)
...
         {
-            foreach (char c in str)
+            if (int.TryParse(str, out int _))
```
Comment (torbacz) [inline nanoFramework.Json/TimeExtensions.cs:132]: Return from TryParse method.

**Comment by torbacz** [comment]
Summary of changes:
- Performance improvements up to 47%!
- Major code refactor (again 😄)
- Replace if/switch types checks with hashtable with converters
- New unit tests for converters 
- Because, now we use converters we can override them in any application or implement new ones for custom conversion 

Latest performance results:
(used this formula to calculate %:  https://www.calculatorsoup.com/calculators/algebra/percentage-increase-calculator.php)
```text
Console export: ReferenceTypesDeserializationBenchmark benchmark class.
| MethodName         | ItterationCount | MeanOld| MeanNew| Improvement
| --------------------------------------------------------------------
| IntArray           | 5               | 260 ms | 146 ms | 44%
| ShortArray         | 5               | 266 ms | 142 ms | 47%
| String             | 5               | 88 ms  | 64 ms  | 27%
| NestedClass        | 5               | 198 ms | 160 ms | 19%
| ComplexObject      | 5               | 334 ms | 252 ms | 25%
| ComplexArrayObject | 5               | 594 ms | 440 ms | 26%
| ArrayList          | 5               | 556 ms | 460 ms | 17%

Console export: ValueTypesDeserializationBenchmark benchmark class.
| MethodName | ItterationCount | MeanOld| MeanNew| Improvement
| ------------------------------------------------------------
| Short      | 5               | 216 ms | 208 ms | 4%
| TimeSpanT  | 5               | 352 ms | 348 ms | 1%
| Float      | 5               | 230 ms | 224 ms | 3%
| Double     | 5               | 242 ms | 238 ms | 2%
| DateTimeT  | 5               | 300 ms | 304 ms | -1%
| Long       | 5               | 242 ms | 238 ms | 2%
| EnumBoxed  | 5               | 388 ms | 318 ms | 18%

Console export: ReferenceTypesSerializationBenchmark benchmark class.
| MethodName                    | ItterationCount | MeanOld | MeanNew | Improvement
| ---------------------------------------------------------------------------------
| IntArray                      | 5               | 788 ms  | 732 ms  | 7%
| ArrayList                     | 5               | 1362 ms | 1192 ms | 12%
| ComplexObjectAzureTwinPayload | 5               | 620 ms  | 572 ms  | 8%
| ShortArray                    | 5               | 836 ms  | 602 ms  | 28%
| String                        | 5               | 180 ms  | 180 ms  | 0%
| NestedClass                   | 5               | 222 ms  | 196 ms  | 12%
| ComplexArrayObject            | 5               | 796 ms  | 734 ms  | 8%

Console export: ValueTypesSerializationBenchmark benchmark class.
| MethodName | ItterationCount | MeanOld | MeanNew | Improvement   
| --------------------------------------------------------------
| Short      | 5               | 222 ms  | 124 ms  | 44%
| TimeSpanT  | 5               | 254 ms  | 138 ms  | 46%
| Float      | 5               | 284 ms  | 176 ms  | 38%
| Double     | 5               | 256 ms  | 164 ms  | 36%
| DateTimeT  | 5               | 1536 ms | 1562 ms | -2%
| Long       | 5               | 224 ms  | 130 ms  | 48%
| Enum       | 5               | 266 ms  | 150 ms  | 44%
| EnumBoxed  | 5               | 932 ms  | 650 ms  | 30%
```

Few more things to do (but not in this PR, it's already huge):
- I'm curious why DateTime is taking sooo long, compared to others
- Some methods in converters never gets hits (e. g. Bool.ToType)
- More cleanup 😄

@josesimoes I think it's ready to be merged.

**Comment by josesimoes** [comment]
@torbacz this is pretty amazing!!

**Comment by Ellerbach** [comment]
@torbacz amazing work! I love the engineering way of doing this! That's great! Maturity of a project is very much related to the phase you're in. When you're about improving performance, you're already in a very advance phase. Especially with great tools and following a great methodology.
We're here now with nanoFramework especially in such fundamental nugets like Json!

**Comment by torbacz** [comment]
@Ellerbach 
Tt's very nice to hear such a compliment from someone who has been working in the industry for so long. Thanks!

**Pull request merged**